src/dataflow-types/src/reconciliation/command.rs

```diff
@@ -0,0 +1,361 @@
+        }
+    }
+
+    async fn observe_command(&mut self, command: Command) -> Result<(), anyhow::Error> {
+        use Command::*;
+        match command {
+            Compute(command, instance) => self.observe_compute_command(command, instance).await,
+            Storage(command) => self.observe_storage_command(command).await,
+        }
+    }
+
+    /// Absorbs a response, and produces response that should be emitted.
+    pub fn absorb_response(&mut self, message: Response) {
+        match message {
+            Response::Compute(ComputeResponse::FrontierUppers(mut list), instance) => {
+                for (id, changes) in list.iter_mut() {
+                    if let Some(frontier) = self.uppers.get_mut(&(*id, Some(instance))) {
+                        let iter = frontier.update_iter(changes.drain());
+                        changes.extend(iter);
+                    } else {
+                        changes.clear();
+                    }
+                }
+
+                self.responses.push_back(Response::Compute(
+                    ComputeResponse::FrontierUppers(list),
+                    instance,
+                ));
+            }
+            // Avoid multiple retractions of minimum time, to present as updates from one worker.
+            Response::Storage(StorageResponse::TimestampBindings(mut feedback)) => {
+                for (id, changes) in feedback.changes.iter_mut() {
+                    if let Some(frontier) = self.uppers.get_mut(&(*id, None)) {
+                        let iter = frontier.update_iter(changes.drain());
+                        changes.extend(iter);
+                    } else {
+                        changes.clear();
+                    }
+                }
+
+                self.responses
+                    .push_back(Response::Storage(StorageResponse::TimestampBindings(
+                        feedback,
+                    )));
+            }
+            Response::Compute(ComputeResponse::PeekResponse(uuid, response), instance) => {
+                self.responses.push_back(Response::Compute(
+                    ComputeResponse::PeekResponse(uuid, response),
+                    instance,
+                ));
+            }
+            Response::Compute(ComputeResponse::TailResponse(id, response), instance) => {
+                self.responses.push_back(Response::Compute(
+                    ComputeResponse::TailResponse(id, response),
+                    instance,
+                ));
+            }
+        }
+    }
+
+    async fn observe_compute_command(
+        &mut self,
+        command: ComputeCommand,
+        instance: ComputeInstanceId,
+    ) -> Result<(), anyhow::Error> {
+        use Command::Compute;
+        use ComputeCommand::*;
+        match command {
+            CreateInstance(config, logging) => {
+                if !self.created {
+                    self.created = true;
```

Comment:
I'm a bit confused now about what this does. We set it for the first create instance, and unset it for any drop?

Reply:
Should `self.created` maybe be a `HashSet<(Option<ComputeInstanceId>)` instead, which is consulted before doing these things, or am I way confused?

Reply:
`created` is now shareded by instance identifier/None for storage.